src/ripple/ledger/impl/View.cpp

```diff
@@ -757,286 +757,77 @@ dirAdd (ApplyView& view,
     std::function<void (SLE::ref)>          fDescriber,
     beast::Journal j)
 {
-    JLOG (j.trace()) << "dirAdd:" <<
-        " dir=" << to_string (dir.key) <<
-        " uLedgerIndex=" << to_string (uLedgerIndex);
+    bool existed = static_cast<bool>(view.peek (dir));
+    bool success;
 
-    auto sleRoot = view.peek(dir);
+    std::tie(success, uNodeDir) = view.dirInsert (
+        dir, uLedgerIndex, true, fDescriber);
 
-    if (! sleRoot)
-    {
-        // No root, make it.
-        sleRoot = std::make_shared<SLE>(dir);
-        sleRoot->setFieldH256 (sfRootIndex, dir.key);
-        view.insert (sleRoot);
-        fDescriber (sleRoot);
-
-        STVector256 v;
-        v.push_back (uLedgerIndex);
-        sleRoot->setFieldV256 (sfIndexes, v);
-
-        JLOG (j.trace()) <<
-            "dirAdd: created root " << to_string (dir.key) <<
-            " for entry " << to_string (uLedgerIndex);
-
-        uNodeDir = 0;
+    if (success)
+        return { tesSUCCESS, !existed };
 
-        return { tesSUCCESS, true };
-    }
-
-    SLE::pointer sleNode;
-    STVector256 svIndexes;
+    return { tecDIR_FULL, false };
+}
 
-    uNodeDir = sleRoot->getFieldU64 (sfIndexPrevious);       // Get index to last directory node.
+static
+boost::optional<std::uint64_t>
+findPage (
```

Comment:
This is called `findPage`, but it only searches the first 20 pages and we have to look at the code to see that. I'm worried someone is going to reuses this and it's going to bite us. Suggestions (any are fine with me):
1) Put the code back inside `dirDelete`, either with a lambda or just embed the code
2) Add a parameter to the function that controls how many pages to search
3) Add a comment documenting that it only searches 20 pages